src/codec/tremor.rs

```diff
@@ -0,0 +1,423 @@
+                w.write_i64::<E>(*n)?;
+            }
+            StaticNode::U64(n) => {
+                w.write_u8(Self::U64)?;
+                w.write_u64::<E>(*n)?;
+            }
+            StaticNode::F64(n) => {
+                w.write_u8(Self::F64)?;
+                w.write_f64::<E>(*n)?;
+            }
+            StaticNode::Bool(true) => w.write_u8(Self::BOOL_TRUE)?,
+            StaticNode::Bool(false) => w.write_u8(Self::BOOL_FALSE)?,
+            StaticNode::Null => w.write_u8(Self::NULL)?,
+        }
+        Ok(())
+    }
+    #[inline]
+    #[allow(clippy::cast_possible_truncation)]
+    fn write_len_type<E: ByteOrder>(t: u8, len: usize, w: &mut impl Write) -> Result<()> {
+        w.write_u8(t)?;
+        w.write_u64::<E>(len as u64)?;
+        Ok(())
+    }
+    #[inline]
+    #[allow(clippy::cast_possible_truncation)]
+    fn write_string<E: ByteOrder>(s: &str, w: &mut impl Write) -> Result<()> {
+        Self::write_len_type::<E>(Self::STRING, s.len(), w)?;
+        w.write_all(s.as_bytes())?;
+        Ok(())
+    }
+    #[inline]
+    #[allow(clippy::cast_possible_truncation)]
+    fn write_bytes<E: ByteOrder>(b: &[u8], w: &mut impl Write) -> Result<()> {
+        Self::write_len_type::<E>(Self::BYTES, b.len(), w)?;
+        w.write_all(b)?;
+        Ok(())
+    }
+    #[inline]
+    #[allow(clippy::cast_possible_truncation)]
+    fn write_array<E: ByteOrder>(a: &[Value], w: &mut impl Write) -> Result<()> {
+        Self::write_len_type::<E>(Self::ARRAY, a.len(), w)?;
+        for v in a {
+            Tremor::encode_::<E>(v, w)?;
+        }
+        Ok(())
+    }
+    #[inline]
+    fn write_object<E: ByteOrder>(
+        o: &HashMap<Cow<'_, str>, Value>,
+        w: &mut impl Write,
+    ) -> Result<()> {
+        Self::write_len_type::<E>(Self::OBJECT, o.len(), w)?;
+        for (k, v) in o.iter() {
+            w.write_u64::<E>(k.len() as u64)?;
+            w.write_all(k.as_bytes())?;
+            Tremor::encode_::<E>(v, w)?;
+        }
+        Ok(())
+    }
+    #[inline]
+    fn encode_<E: ByteOrder>(value: &Value, w: &mut impl Write) -> Result<()> {
+        match value {
+            Value::Static(s) => Tremor::write_static::<E>(s, w),
+            Value::String(s) => Self::write_string::<E>(s, w),
+            Value::Array(a) => Self::write_array::<E>(a, w),
+            Value::Object(o) => Self::write_object::<E>(o.as_ref(), w),
+            Value::Bytes(b) => Self::write_bytes::<E>(b, w),
+        }
+    }
+}
+
+impl Codec for Tremor {
+    fn name(&self) -> &str {
+        "tremor"
+    }
+
+    fn mime_types(&self) -> Vec<&'static str> {
+        vec!["application/octet-stream"]
+    }
+
+    fn decode<'input>(
+        &mut self,
+        data: &'input mut [u8],
+        _ingest_ns: u64,
+    ) -> Result<Option<Value<'input>>> {
+        Tremor::decode(data).map(Some)
+    }
+
+    fn encode(&self, data: &Value) -> Result<Vec<u8>> {
+        let mut dst = Vec::with_capacity(512);
```

Comment:
this could be made configurable, so people could optimize for their expected sizes.

Reply:
We don't have a way to make codecs configurable, I like the idea, I would love to generalize this as we have a few codecs that have the same logic, but I don't think  this is the right PR to "make codecs configurable". I'll ticket it!

Reply:
fild under #2130

Reply:
Codecs are configurable. `resolve` takes a `NameWithConfig`.
It is just that none of the existing codecs is doing it as of now.
Yours could be the first.

For inspiration on how to do it look into the preprocessors for example :)

Reply:
Nice, so I did change the whole thing around to avoid the problem and let the codec remember what the sizes are and allocate to fit the largest so it'll armotize against a single allocation w/o user interaction now :D

Reply:
We came to a solution on this so I'm closing it :)